Library/OcCryptoLib/RsaDigitalSign.c

```diff
@@ -0,0 +1,511 @@
+  Exponent depends on the configuration (65537 (default), or 3).
+
+  @param Key        Key to use in signing
+  @param InOut      Input and output big-endian byte array
+ **/
+STATIC
+VOID
+ModPow (
+  RSA_PUBLIC_KEY  *Key,
+  UINT8           *InOut
+  )
+{
+  UINT32 *A     = NULL;
+  UINT32 *Ar    = NULL;
+  UINT32 *Aar   = NULL;
+  UINT32 *Aaa   = NULL;
+  INT32  Index  = 0;
+  UINT32 Tmp    = 0;
+
+  A = AllocateZeroPool (Key->Size * sizeof (UINT32));
+  Ar = AllocateZeroPool (Key->Size * sizeof (UINT32));
+  Aar = AllocateZeroPool (Key->Size * sizeof (UINT32));
+  if (A == NULL || Ar == NULL || Aar == NULL) {
+    if (A != NULL) {
+      FreePool (A);
+    }
+    if (Ar != NULL) {
+      FreePool (Ar);
+    }
+    if (Aar != NULL) {
+      FreePool (Aar);
+      // TODO: debug_error about efi memory alocation
+    }
+    return;
+  }
+
+  //
+  // Re-use location
+  //
+  Aaa = Aar;
+
+  //
+  // Convert from big endian byte array to little endian word array
+  //
+  for (Index = 0; Index < (INT32) Key->Size; Index++) {
+    Tmp =
+      ((UINT32)InOut[((Key->Size - 1 - Index) * 4) + 0] << 24) |
+      ((UINT32)InOut[((Key->Size - 1 - Index) * 4) + 1] << 16) |
+      ((UINT32)InOut[((Key->Size - 1 - Index) * 4) + 2] << 8) |
+      ((UINT32)InOut[((Key->Size - 1 - Index) * 4) + 3] << 0);
+    A[Index] = Tmp;
+  }
+
+  //
+  // Ar = A * Rr / R mod M
+  //
+  MontMul (Key, Ar, A, Key->Rr);
+  //
+  // Exponent 65537
+  //
+  for (Index = 0; Index < 16; Index += 2) {
+    //
+    // Aar = Ar * Ar / R mod M 
+    //
+    MontMul (Key, Aar, Ar, Ar);
+    //
+    // Ar = Aar * Aar / R mod M 
+    //
+    MontMul (Key, Ar, Aar, Aar);
+  }
+  //
+  // Aaa = Ar * A / R mod M
+  //
+  MontMul (Key, Aaa, Ar, A);
+
+  //
+  // Make sure Aaa < Mod; Aaa is at most 1x mod too large.
+  //
+  if (GeMod (Key, Aaa)){
+    SubMod (Key, Aaa);
+  }
+
+  //
+  // Convert to bigendian byte array
+  //
+  for (Index = (INT32) Key->Size - 1; Index >= 0; --Index) {
+    Tmp = Aaa[Index];
+
+    *InOut++ = (UINT8) (Tmp >> 24);
+    *InOut++ = (UINT8) (Tmp >> 16);
+    *InOut++ = (UINT8) (Tmp >>  8);
+    *InOut++ = (UINT8) (Tmp >>  0);
+  }
+
+  //
+  // Free work buffers before return
+  //
+  if (A != NULL) {
+    FreePool (A);
+  }
+  if (Ar != NULL) {
+    FreePool (Ar);
+  }
+  if (Aar != NULL) {
+    FreePool (Aar);
+  }  
+}
+
+/**
+  This routine parses key data from RsaPublicKey Header
+
+  @ param KeyData         RSA public key data
+  @ param Length          RSA public key size
+ */
+STATIC
+RSA_PUBLIC_KEY
+*RsaParseKeyData (
+  UINT8  *KeyData,
+  UINTN  Length
+  )
+{
+  RSA_PUBLIC_KEY_HEADER  KeyHeader;
+  CONST UINT8            *N;
+  CONST UINT8            *Rr;
+  UINT32                 Index           = 0;
+  UINT32                 RsaKeyStructLen = 0;
+  RSA_PUBLIC_KEY         *RsaPublicKey   = NULL;
+  UINTN                  ExpectedLength  = 0;
+
+  //
+  // Copy KeyHeader
+  //
+  CopyMem (&KeyHeader, KeyData, sizeof (RSA_PUBLIC_KEY_HEADER));
```

Comment:
This assumes KeyData starts with RSA_PUBLIC_KEY_HEADER, however KeyData is UINT8 *, which to me indicates an array or an arbitrary buffer. Is there any specific reason KeyData is UINT8 * and not RSA_PUBLIC_KEY_HEADER *?

Reply:
I would ASSERT on minimum length before actually accessing the header

Reply:
We should validate KeyData and check byte ordering before we assign RSA_PUBLIC_KEY_HEADER type to this pointer. Currently it is not implemented and replaced by dummy CopyMem.